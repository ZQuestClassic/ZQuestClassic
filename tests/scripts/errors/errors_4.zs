#option NO_ERROR_HALT on

class Object {
}

global script Global
{
    void run()
    {
		Object obj = new Object();
		int numbers[] = {1, 2, 3};

		// Invalid.
		Object n1 = ArrayPopBack(numbers);
		ArrayPushBack(numbers, obj);
		ArrayPushBack(numbers, "");
		ArrayPushBack(numbers, 1L); // TODO: this should be invalid (else: we perform an implicit cast and multiply by 10000...)
		Object choice_1 = Choose(1, 2, 3);
		auto choice_2 = Choose(1, 2, obj);
		auto max_1 = Max(1, 2, obj);
		auto max_2 = Max(1, obj);
		auto max_3 = Max(1);

		SizeOfArray(Hero); // TODO: this should warn since the type was coerced to an array (LEGACY_ARRAYS).

		// OK.
		Object objects2[] = {new Object()};
		Object o = ArrayPopBack(objects2);
		int n2 = ArrayPopBack(numbers);
		ArrayPushBack(numbers, 1);
		ArrayPushBack(numbers, <untyped>obj);
		int choice_3 = Choose(1, 2, 3);
		auto choice_4 = Choose(1, 2, 3);
		auto max_2 = Max(1, 2);
		max_2 = Max(1, 2, 3);
		max_2 = Max(1, 2, 3, Max(3, 4));
    }

	void old_ptrs(char32 ptr)
	{
		// For compatability with old scripts, the template deduction code allows
		// matching char32 as an array type by default.

		// OK
		ArrayPushBack(ptr, 1);
		ResizeArray(ptr, 1);

		// OK
		{
			#option LEGACY_ARRAYS on

			ArrayPushBack(ptr, 1);
			ResizeArray(ptr, 1);
		}

		// Errors
		{
			#option LEGACY_ARRAYS off

			ArrayPushBack(ptr, 1);
			ResizeArray(ptr, 1);
		}
	}
}
