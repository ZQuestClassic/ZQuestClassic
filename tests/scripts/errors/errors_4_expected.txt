stderr:

Compiling 'errors_4.zs'
Pass 1: Parsing
Pass 2: Preprocessing
Pass 3: Registration
Pass 4: Analyzing Code
errors_4.zs Line 60 @ Columns 4-17 - Error T021: Function ArrayPushBack(char32, int) has not been declared.

58                #option LEGACY_ARRAYS off
59    
60                ArrayPushBack(ptr, 1);
                  ^~~~~~~~~~~~~


errors_4.zs Line 61 @ Columns 4-15 - Error T021: Function ResizeArray(char32, int) has not been declared.

59    
60                ArrayPushBack(ptr, 1);
61                ResizeArray(ptr, 1);
                  ^~~~~~~~~~~


errors_4.zs Line 14 @ Columns 10-12 - Error T017: Cannot cast from int to Object.

12    
13            // Invalid.
14            Object n1 = ArrayPopBack(numbers);
                     ^~


errors_4.zs Line 15 @ Columns 3-16 - Error T021: Function ArrayPushBack(int[], Object) has not been declared.

13    // Invalid.
14    Object n1 = ArrayPopBack(numbers);
15    ArrayPushBack(numbers, obj);
      ^~~~~~~~~~~~~


errors_4.zs Line 16 @ Columns 3-16 - Error T021: Function ArrayPushBack(int[], char32[]) has not been declared.

14    Object n1 = ArrayPopBack(numbers);
15    ArrayPushBack(numbers, obj);
16    ArrayPushBack(numbers, "");
      ^~~~~~~~~~~~~


errors_4.zs Line 18 @ Columns 10-18 - Error T017: Cannot cast from int to Object.

16    ArrayPushBack(numbers, "");
17    ArrayPushBack(numbers, 1L); // TODO: this should be invalid (else: we perform an implicit cast and multiply by 10000...)
18    Object choice_1 = Choose(1, 2, 3);
             ^~~~~~~~


errors_4.zs Line 19 @ Columns 19-25 - Error T021: Function Choose(int, int, Object) has not been declared.

17    ArrayPushBack(numbers, 1L); // TODO: this should be invalid (else: we perform an implicit cast and multiply by 10000...)
18    Object choice_1 = Choose(1, 2, 3);
19    auto choice_2 = Choose(1, 2, obj);
                      ^~~~~~


errors_4.zs Line 20 @ Columns 16-19 - Error T021: Function Max(int, int, Object) has not been declared.

18    Object choice_1 = Choose(1, 2, 3);
19    auto choice_2 = Choose(1, 2, obj);
20    auto max_1 = Max(1, 2, obj);
                   ^~~


errors_4.zs Line 21 @ Columns 16-19 - Error T021: Function Max(int, Object) has not been declared.

19    auto choice_2 = Choose(1, 2, obj);
20    auto max_1 = Max(1, 2, obj);
21    auto max_2 = Max(1, obj);
                   ^~~


errors_4.zs Line 22 @ Columns 16-19 - Error T021: Function Max(int) has not been declared.

20    auto max_1 = Max(1, 2, obj);
21    auto max_2 = Max(1, obj);
22    auto max_3 = Max(1);
                   ^~~


Failure!
Compile finished with exit code '21' (compiled with errors)

stdout:

{
  "success": false,
  "code": 21,
  "diagnostics": [
    {
      "range": {
        "start": {
          "line": 59,
          "character": 3
        },
        "end": {
          "line": 59,
          "character": 16
        }
      },
      "severity": 1,
      "message": "T021: Function ArrayPushBack(char32, int) has not been declared."
    },
    {
      "range": {
        "start": {
          "line": 60,
          "character": 3
        },
        "end": {
          "line": 60,
          "character": 14
        }
      },
      "severity": 1,
      "message": "T021: Function ResizeArray(char32, int) has not been declared."
    },
    {
      "range": {
        "start": {
          "line": 13,
          "character": 9
        },
        "end": {
          "line": 13,
          "character": 11
        }
      },
      "severity": 1,
      "message": "T017: Cannot cast from int to Object."
    },
    {
      "range": {
        "start": {
          "line": 14,
          "character": 2
        },
        "end": {
          "line": 14,
          "character": 15
        }
      },
      "severity": 1,
      "message": "T021: Function ArrayPushBack(int[], Object) has not been declared."
    },
    {
      "range": {
        "start": {
          "line": 15,
          "character": 2
        },
        "end": {
          "line": 15,
          "character": 15
        }
      },
      "severity": 1,
      "message": "T021: Function ArrayPushBack(int[], char32[]) has not been declared."
    },
    {
      "range": {
        "start": {
          "line": 17,
          "character": 9
        },
        "end": {
          "line": 17,
          "character": 17
        }
      },
      "severity": 1,
      "message": "T017: Cannot cast from int to Object."
    },
    {
      "range": {
        "start": {
          "line": 18,
          "character": 18
        },
        "end": {
          "line": 18,
          "character": 24
        }
      },
      "severity": 1,
      "message": "T021: Function Choose(int, int, Object) has not been declared."
    },
    {
      "range": {
        "start": {
          "line": 19,
          "character": 15
        },
        "end": {
          "line": 19,
          "character": 18
        }
      },
      "severity": 1,
      "message": "T021: Function Max(int, int, Object) has not been declared."
    },
    {
      "range": {
        "start": {
          "line": 20,
          "character": 15
        },
        "end": {
          "line": 20,
          "character": 18
        }
      },
      "severity": 1,
      "message": "T021: Function Max(int, Object) has not been declared."
    },
    {
      "range": {
        "start": {
          "line": 21,
          "character": 15
        },
        "end": {
          "line": 21,
          "character": 18
        }
      },
      "severity": 1,
      "message": "T021: Function Max(int) has not been declared."
    }
  ],
  "metadata": {
    "currentFileSymbols": 18,
    "symbols": 18,
    "identifiers": 47,
    "elided": true
  }
}
